python_modules/libraries/dagster-dg/dagster_dg/docs.py

```diff
@@ -0,0 +1,168 @@
+import tempfile
+import webbrowser
+from dataclasses import dataclass
+from typing import AbstractSet, Any, Mapping, Optional, Sequence, Union
+
+import markdown
+import yaml
+
+from dagster_dg.context import RemoteComponentType
+
+REF_BASE = "#/$defs/"
+COMMENTED_MAPPING_TAG = "!commented_mapping"
+
+
+@dataclass(frozen=True)
+class CommentedObject:
+    key: str
+    value: Union[Sequence["CommentedObject"], Any]
+    comment: Optional[str]
+    top_level: bool = False
+
+
+def _dereference_schema(
+    json_schema: Mapping[str, Any], subschema: Mapping[str, Any]
+) -> Mapping[str, Any]:
+    if "$ref" in subschema:
+        return json_schema["$defs"].get(subschema["$ref"][len(REF_BASE) :])
+    else:
+        return subschema
+
+
+def _get_commented_obj(
+    name: str,
+    json_schema: Mapping[str, Any],
+    subschema: Mapping[str, Any],
+    available_scope: Optional[AbstractSet[str]] = None,
+) -> Union[CommentedObject, Any]:
+    additional_scope = subschema.get("dagster_available_scope")
+    available_scope = (available_scope or set()) | set(additional_scope or [])
+
+    subschema = _dereference_schema(json_schema, subschema)
+    if "anyOf" in subschema:
+        return _get_commented_obj(
+            name, json_schema, subschema["anyOf"][0], available_scope=available_scope
+        )
+
+    objtype = subschema["type"]
+    if objtype == "object":
+        return CommentedObject(
+            key=name,
+            value={
+                k: _get_commented_obj(k, json_schema, v)
+                for k, v in subschema.get("properties", {}).items()
+            },
+            comment=f"Available scope: {available_scope}" if available_scope else None,
+        )
+    elif objtype == "array":
+        return [
+            _get_commented_obj(
+                name, json_schema, subschema["items"], available_scope=available_scope
+            )
+        ]
+    elif objtype == "string":
+        return "..."
+    elif objtype == "integer":
+        return 0
+    elif objtype == "boolean":
+        return False
+    else:
+        return f"<{objtype}>"
+
+
+class ComponentDumper(yaml.SafeDumper):
+    def increase_indent(self, flow=False, *args, **kwargs):
+        return super().increase_indent(flow=flow, indentless=False)
+
+    def write_line_break(self) -> None:
+        # add an extra line break between top-level keys
+        if self.indent == 0:
+            super().write_line_break()
+        super().write_line_break()
+
+    def _get_tag(self) -> str:
+        return getattr(self.event, "tag", "")
+
+    def expect_node(self, root=False, sequence=False, mapping=False, simple_key=False):
+        # for commented mappings, emit comment above the value
+        tag = self._get_tag()
```

Comment:
It's painful that we have to do this, but this impl doesn't feel too hacky to me